packages/contracts-bedrock/src/L1/OPContractsManager.sol

```diff
@@ -1174,6 +1204,276 @@ contract OPContractsManagerDeployer is OPContractsManagerBase {
+    ///         the caller. Other parameters are selected automatically.
+    struct GameParameters {
+        address proposer;
+        address challenger;
+        uint256 maxGameDepth;
+        uint256 splitDepth;
+        uint256 initBond;
+        Duration clockExtension;
+        Duration maxClockDuration;
+    }
+
+    /// @notice Input parameters for the migration.
+    struct MigrateInput {
+        bool usePermissionlessGame;
+        Proposal startingAnchorRoot;
+        GameParameters gameParameters;
+        OPContractsManager.OpChainConfig[] opChainConfigs;
+    }
+
+    /// @param _contractsContainer Container of blueprints and implementations.
+    constructor(OPContractsManagerContractsContainer _contractsContainer) OPContractsManagerBase(_contractsContainer) { }
+
+    /// @notice Migrates one or more OP Stack chains to use the Super Root dispute games and shared
+    ///         dispute game contracts.
+    /// @param _input The input parameters for the migration.
+    function migrate(MigrateInput calldata _input) public virtual {
+        // Check that all of the configs have the same proxy admin owner and prestate.
+        for (uint256 i = 0; i < _input.opChainConfigs.length; i++) {
+            // Different chains might actually have different ProxyAdmin contracts, but it's fine
+            // as long as the owner of all of those contracts is the same.
+            if (_input.opChainConfigs[i].proxyAdmin.owner() != _input.opChainConfigs[0].proxyAdmin.owner()) {
+                revert OPContractsManagerInteropMigrator_ProxyAdminOwnerMismatch();
+            }
+            if (_input.opChainConfigs[i].absolutePrestate.raw() != _input.opChainConfigs[0].absolutePrestate.raw()) {
+                revert OPContractsManagerInteropMigrator_AbsolutePrestateMismatch();
+            }
+        }
+
+        // Grab an array of portals from the configs.
+        IOptimismPortal[] memory portals = new IOptimismPortal[](_input.opChainConfigs.length);
+        for (uint256 i = 0; i < _input.opChainConfigs.length; i++) {
+            portals[i] = IOptimismPortal(payable(_input.opChainConfigs[i].systemConfigProxy.optimismPortal()));
+        }
+
+        // Check that the portals have the same SuperchainConfig.
+        for (uint256 i = 0; i < portals.length; i++) {
+            if (portals[i].superchainConfig() != portals[0].superchainConfig()) {
+                revert OPContractsManagerInteropMigrator_SuperchainConfigMismatch();
+            }
+        }
+
+        // NOTE that here and in the rest of this function, we are using the first provided chain's
+        // ProxyAdmin contract as the ProxyAdmin for all of the newly shared contracts. This is
+        // safe because we already checked that all of the provided chains have the same ProxyAdmin
+        // owner and therefore have the same access models.
+        address proxyAdminOwner = _input.opChainConfigs[0].proxyAdmin.owner();
```

Comment:
Comment only: It would be nice if we could easily do this, but I see that it results in stack too deep.
```suggestion
        IProxyAdmin proxyAdmin = _input.opChainConfigs[0].proxyAdmin;
        address proxyAdminOwner = _input.opChainConfigs[0].proxyAdmin.owner();
```

Reply:
Yeah all the ugliness was to avoid stack-too-deep here